universum/configuration_support.py

```diff
@@ -117,57 +421,73 @@ class Variations(list):
         then all dictionaries from `other`.
 
-        :param other: `Variations` object OR list to be added to `self`
+        :param other: `Variations` object OR list of project configurations to be added to `configs`
         :return: new `Variations` object, including all configurations from both `self` and `other` objects
         """
-        return Variations(list.__add__(list(self), other))
+        list_other: List[ProjectConfiguration] = other.configs if isinstance(other, Variations) else other
+        return Variations(list.__add__(self.configs, list_other))
 
-    def __mul__(self, other: Union[int, 'Variations']) -> 'Variations':
+    def __mul__(self, other: Union['Variations', int]) -> 'Variations':
         """
         This functions defines operator ``*`` for :class:`.Variations` class objects.
-        The resulting object is created by combining every `self` list member with
+        The resulting object is created by combining every `configs` list member with
         every `other` list member using :func:`.combine()` function.
 
         :param other: `Variations` object  OR an integer value to be multiplied to `self`
         :return: new `Variations` object, consisting of the list of combined configurations
         """
 
-        if isinstance(other, six.integer_types):
-            result_list: List = list.__mul__(list(self), other)
-        else:
-            result_list = []
-            for obj_a in self:
-                obj_a_copy = copy.deepcopy(obj_a)
-                if "children" in obj_a:
-                    obj_a_copy["children"] = obj_a["children"] * other
-                else:
-                    obj_a_copy["children"] = copy.deepcopy(other)
-                obj_a_copy["skip_numbering_level"] = len(self) <= 1
-
-                result_list.append(obj_a_copy)
+        if isinstance(other, int):
+            return Variations(list.__mul__(self.configs, other))
+        config_list: List[ProjectConfiguration] = []
+        for obj_a in self.configs:
+            obj_a_copy = copy.deepcopy(obj_a)
+            if obj_a.children:
+                obj_a_copy.children = obj_a.children * other
+            else:
+                obj_a_copy.children = copy.deepcopy(other)
+            obj_a_copy.skip_numbering_level = len(self.configs) <= 1
 
-        result = Variations(result_list)
-        return result
+            config_list.append(obj_a_copy)
+        return Variations(config_list)
 
-    def all(self) -> Iterable[Dict]:
+    def all(self) -> Iterable[ProjectConfiguration]:
         """
         Function for configuration iterating.
 
         :return: iterable for all dictionary objects in :class:`.Variations` list
         """
-        for obj_a in self:
+        for obj_a in self.configs:
 
             obj_a_copy = copy.deepcopy(obj_a)
 
-            if "children" in obj_a_copy:
-                for obj_b in obj_a_copy["children"].all():
+            if obj_a_copy.children:
```

Comment:
Is it equivalent change?

Reply:
it's not equivalent, but I think this change is warranted: previously configuration could lack some fields, and universum had lots of 'get' calls with provided default values - this was risky in case default was incorrectly set to 2 different values in 2 different parts of code. Instead ProjectConfiguration is constructed with all properties set, but all of them are 'falsy' by default. This means that there's an internal guarantee of returned type (no 'Optional'), which simplifies the code.

Reply:
Ok, just to make it clear: are you saying that obj_a_copy is guaranteed to have "children" field?
If it is guaranteed by the static typing, I'm ok with the change.

Reply:
children is a property so it's guaranteed, but it can return None which is falsy